sigstore/_internal/trust.py

```diff
@@ -355,55 +372,73 @@ def from_file(
         inner = _SigningConfig().from_json(Path(path).read_bytes())
         return cls(inner)
 
-    @staticmethod
-    def _get_valid_service_url(services: list[Service]) -> str | None:
+    def _get_valid_services(
+        self,
+        services: list[Service],
+        valid_versions: list[int],
+        config: ServiceConfiguration | None,
+    ) -> list[Service]:
+        """Return supported services, taking SigningConfig restrictions into account"""
+
+        # split logs by operator, only include valid services
+        logs_by_operator: dict[str, list[Service]] = defaultdict(list)
         for service in services:
-            if service.major_api_version != 1:
+            if service.major_api_version not in valid_versions:
                 continue
 
             if not _is_timerange_valid(service.valid_for, allow_expired=False):
                 continue
-            return service.url
-        return None
 
-    def get_tlog_urls(self) -> list[str]:
+            logs_by_operator[service.operator].append(service)
+
+        # build a list of services but make sure we only include logs of one version per operator
+        result: list[Service] = []
+        for logs in logs_by_operator.values():
+            logs.sort(key=lambda s: -s.major_api_version)
+            max_version = logs[-1].major_api_version
+
+            while logs and logs[-1].major_api_version == max_version:
+                result.append(logs.pop())
+
+        # limit the list based on ServiceConfiguration
+        if not config or config.selector == ServiceSelector.ALL:
+            return result
+
+        count = config.count if config.selector == ServiceSelector.EXACT else 1
```

Comment:
Add a comment saying this for when the selector is ANY or undefined.

Reply:
I suppose I should handle those separately since undefined should be an error ... I really don't get why it needs to be there...